src/librustc_middle/mir/mod.rs

```diff
@@ -72,15 +72,21 @@ impl<'tcx> HasLocalDecls<'tcx> for Body<'tcx> {
 
 /// The various "big phases" that MIR goes through.
 ///
+/// These phases all describe dialects of MIR. Since all MIR uses the same datastructures, the
+/// dialects forbid certain variants or values in certain phases.
+///
 /// Warning: ordering of variants is significant.
 #[derive(Copy, Clone, RustcEncodable, RustcDecodable, Debug, PartialEq, Eq, PartialOrd, Ord)]
 #[derive(HashStable)]
 pub enum MirPhase {
     Build = 0,
     Const = 1,
     Validated = 2,
+    /// After this phase, the only `AggregateKind`s allowed are `Array` and `Generator`
     DropElab = 3,
```

Comment:
"after" this phase? I'd expect comments explaining what happens *in* this phase.

In fact, this is somewhat confusing... MIR validation needs to know the "MIR dialect" being used, and phases are the thing that translates from one MIR dialect to another. So we are using the *transitions* here to name the *states* for validation.

Reply:
I was worried that adding new phases without updating the comments would make them wrong if e.g. a new phase were added right after `DropElab`, but the comment from the current phase 4 would not have been updated.

Reply:
Maybe it would be better to have an explicit `MirDialect` enum, and then each phase is defined as translating from one dialect to another (or the same).

Reply:
Hmm... we'd still have a dialect ordering, so we'd end up duplicating elements from the phase enum (though we'd merge a few like `Build` and `Const`). While I like the explicitness, I also like the simplicity of using phases for dialects.

Reply:
I wouldn't call it a duplicate. In my head I see a directed graph where dialects are nodes and phases are edges. Of course every edge goes to some node, but that's not a duplication of information, that's interesting structure.

> I also like the simplicity of using phases for dialects.

I don't think this is simple at all. To understand your code I have to translate to the graph I mentioned above in my head, and then translate back. "Why does validation still allow DropAndReplace for DropElab? Ah right this is because we are checking the node *before* the DropElab phase."

If you prefer the current state, I can live with it, but know that it is very far from simple IMO.^^

Reply:
> Why does validation still allow DropAndReplace for DropElab? Ah right this is because we are checking the node _before_ the DropElab phase

I think this is why I didn't understand why you found the current scheme confusing. To me it was "DropElab gets rid of `DropAndReplace`, thus only phases after it check that it's gone". Even if there are parts in the `DropElab` that would not have `DropAndReplace` anymore, the dialect only officially changes after the phase is finished.

Reply:
Right, but this is only indirectly visible in the code. Can we write things in validation like `if phase > DropElab { no more DropAndReplace }`? Then the phase name that makes the difference is right there, and all my confusing is condensed into the fact that we have a `>` and not a `>=`.

Reply:
There are no more `>=` anyhwere. It's either `>` or an exhaustive match.